### DISCUSSION TITLE
iPXE booting to UEFI drops USB keyboard support on select systems only

### DISCUSSION THREAD

**Discussion by JBWinnipeg** [ipxe/ipxe#308] [Q&A] [answered]
Greetings,

Having an issue specifically with UEFI booting to iPXE IPv4.

For many of our systems, booting using either UEFI or Legacy IPv4 > iPXE > iPXE Boot Menu > Choice to OS deployment works without a problem.  We've had this system in place for months now and it's only come to my attention in the past week when we needed to re-image and re-deploy a few workstations that were deployed some months or a year+ ago.

With Dell Optiplex 5050s however, using UEFI pxe >  iPv4 > iPXE >iPXE Boot Menu *USB input from keyboard is dropped at this point* and without plugging in a ps2 keyboard, we stall out here.  

If I plug in a ps2 keyboard, we can continue the process and the imaging works but I don't know what's happening specifically on the 5050s that the USB keyboard input is dropped and we're unable to choose a menu item to continue the process.

We're using iPXE 1.21.1 (g1192e) 

I've double and triple checked all the BIOS settings for what little legacy USB support remains in the bios, all seems to pertain specifically to booting storage devices and not actual base level connectivity. 

POST behavior > FastBoot is set to 'thorough' but on 'minimal' or 'Auto' the behaviour doesn't change.

Curious if anyone else has experienced this before or has any suggestions on what I could try to resolve this?

**Comment by sebaxakerhtc** [comment, marked as answer]
Hi.
Did you try to add uefi usb at usb.h?
It very slow, but maybe it will work.
`//#define	USB_HCD_USBIO	/* Very slow EFI USB host controller */`

**Reply by JBWinnipeg** [reply]
I wasn't the one in charge of setting up the iPXE config but WONDERFULLY this is exactly what the problem was!  Adding that usb controller definition has resolved the issue!

Thanks!

**Reply by sebaxakerhtc** [reply]
Very nice!
Glad to know it helps 👍

**Reply by EslamE1Torky** [reply]
> Hi. Did you try to add uefi usb at usb.h? It very slow, but maybe it will work. `//#define USB_HCD_USBIO /* Very slow EFI USB host controller */`

can you guys help me where to add the usb controller definition on which file and thanks

**Reply by sebaxakerhtc** [reply]
> > Hi. Did you try to add uefi usb at usb.h? It very slow, but maybe it will work. `//#define USB_HCD_USBIO /* Very slow EFI USB host controller */`
> 
> can you guys help me where to add the usb controller definition on which file and thanks

Inside the https://github.com/ipxe/ipxe/blob/master/src/config/usb.h file
find this
```
//#define	USB_HCD_USBIO	/* Very slow EFI USB host controller */
```
and uncomment it like this
```
#define	USB_HCD_USBIO	/* Very slow EFI USB host controller */
```

**Comment by mcb30** [comment]
Which iPXE binary are you using, and what configuration changes have you made?
